File: cpp/src/cas_counter.cpp
// Compares a manual CAS loop against std::atomic<int>::fetch_add,
// and shows off relaxed ordering -- something Java's atomics never
// expose (java.util.concurrent.atomic is always effectively seq_cst).
// See include/cas_counter.hpp for manualCasIncrement itself (extracted
// there so tests/cas_counter_test.cpp can exercise it directly).

#include "cas_counter.hpp"
#include <thread>
#include <vector>
#include <iostream>
#include <chrono>

constexpr int THREADS = 8;
constexpr int INCREMENTS_PER_THREAD = 200'000;

template <typename Fn>
long long runAndTime(const char* label, Fn task) {
    std::vector<std::thread> threads;
    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < THREADS; ++i) threads.emplace_back(task);
    for (auto& t : threads) t.join();
    auto elapsedMs = std::chrono::duration_cast<std::chrono::milliseconds>(
                         std::chrono::steady_clock::now() - start).count();
    std::cout << label << " took " << elapsedMs << " ms\n";
    return elapsedMs;
}

int main() {
    std::atomic<int> manual{0};
    std::atomic<int> builtin{0};

    runAndTime("manual CAS loop (relaxed)      ", [&] {
        for (int i = 0; i < INCREMENTS_PER_THREAD; ++i) manualCasIncrement(manual);
    });
    runAndTime("fetch_add (relaxed)            ", [&] {
        for (int i = 0; i < INCREMENTS_PER_THREAD; ++i)
            builtin.fetch_add(1, std::memory_order_relaxed);
    });

    std::cout << "manual  = " << manual.load() << "\n";
    std::cout << "builtin = " << builtin.load() << "\n";
    return 0;
}

File: java/src/CasCounter.java
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.LongAdder;

/**
 * Compares three ways to count concurrently in Java:
 *  1) A manual CAS loop over AtomicInteger (shows the mechanics).
 *  2) AtomicInteger.incrementAndGet (same thing, built in).
 *  3) LongAdder, which stripes the counter across cache lines to
 *     reduce CAS contention under heavy parallelism.
 *
 * All three give the same *correctness* guarantee (java.util.concurrent
 * atomics always use a seq-cst-like ordering -- there's no relaxed dial),
 * but LongAdder trades a little read-side cost for far better write
 * throughput when many threads increment at once.
 */
public class CasCounter {

    private static final int THREADS = 8;
    private static final int INCREMENTS_PER_THREAD = 200_000;

    // (1) Manual CAS loop -- illustrates what incrementAndGet does internally.
    static int manualCasIncrement(AtomicInteger counter) {
        int prev, next;
        do {
            prev = counter.get();
            next = prev + 1;
        } while (!counter.compareAndSet(prev, next));
        return next;
    }

    public static void main(String[] args) throws InterruptedException {
        AtomicInteger manual = new AtomicInteger(0);
        AtomicInteger builtin = new AtomicInteger(0);
        LongAdder adder = new LongAdder();

        Runnable manualTask = () -> {
            for (int i = 0; i < INCREMENTS_PER_THREAD; i++) manualCasIncrement(manual);
        };
        Runnable builtinTask = () -> {
            for (int i = 0; i < INCREMENTS_PER_THREAD; i++) builtin.incrementAndGet();
        };
        Runnable adderTask = () -> {
            for (int i = 0; i < INCREMENTS_PER_THREAD; i++) adder.increment();
        };

        runAndTime("manual CAS loop", manualTask);
        runAndTime("AtomicInteger.incrementAndGet", builtinTask);
        runAndTime("LongAdder.increment", adderTask);

        System.out.println("manual  = " + manual.get());
        System.out.println("builtin = " + builtin.get());
        System.out.println("adder   = " + adder.sum());
    }

    private static void runAndTime(String label, Runnable task) throws InterruptedException {
        Thread[] threads = new Thread[THREADS];
        long start = System.nanoTime();
        for (int i = 0; i < THREADS; i++) {
            threads[i] = new Thread(task);
            threads[i].start();
        }
        for (Thread t : threads) t.join();
        long elapsedMs = (System.nanoTime() - start) / 1_000_000;
        System.out.printf("%-32s took %d ms%n", label, elapsedMs);
    }
}
